flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/AndFilterFunction.java

```diff
@@ -1,55 +0,0 @@
-/*
- * Licensed to the Apache Software Foundation (ASF) under one
- * or more contributor license agreements.  See the NOTICE file
- * distributed with this work for additional information
- * regarding copyright ownership.  The ASF licenses this file
- * to you under the Apache License, Version 2.0 (the
- * "License"); you may not use this file except in compliance
- * with the License.  You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-package org.apache.flink.cep.pattern;
-
-import org.apache.flink.api.common.functions.FilterFunction;
-
-/**
- * A filter function which combines two filter functions with a logical and. Thus, the filter
- * function only returns true, iff both filters return true.
- *
- * @param <T> Type of the element to filter
- * @deprecated This is only used when migrating from an older Flink version.
- * Use the {@link org.apache.flink.cep.pattern.conditions.AndCondition} instead.
- */
-@Deprecated
-public class AndFilterFunction<T> implements FilterFunction<T> {
```

Comment:
I'm kind of worried that migration test doesn't cover for and/or/subtype filters/conditions. Could you add a test which proves that migration from 1.3 with such expressions is still working?

Reply:
But it was not possible to have those `*FilterFunction` in code in 1.3. There is no API that uses them. The only reason for those classes was migrating from 1.2 and as we drop that feature in this PR those classes are useless. After deserialization those functions were migrated to `*Condition`.

If I wanted to test the migration that uses the `*FilterFunction` I would need to test migration from 1.2, which was dropped.

Or did I misunderstand you?

Reply:
Sorry
> I'm kind of worried 

I meant: "it seems to me like this change should be ok, the only thing that I'm not 100% sure". In other words I don't expect to have find any issues with this code basing on my understanding of it.

The case and the test that I had in mind should make sure, that checkpoints created/taken in 1.3.x do not use those `*FilterFunction` classes. Correct me if I'm wrong, but I have an impression if 1.3.x somehow created a checkpoint using `*FilterFunction`(unlikely? impossible?), before your change such checkpoint would still correctly load. That's why I'm missing those `and`/`or` conditions in `CEPMigrationTest` from `1.3.x` -> `1.4+`. Also such test might be useful for the future.

Maybe with some extra knowledge about the previous changes this code has undergone it's obvious that this code just works, but I'm missing this context (I'm new to the CEP stuff :( )

Reply:
I am afraid your worries are justified :( 

I analyzed the code once again and indeed it is possible to have a checkpoint taken in 1.3.x that has serialized  `*FilterFunction` classes through the `FilterWrapper` class. It is possible when the job was previously restored from 1.2.x checkpoint. Unfortunately I will need to restore those classes. I will do that tomorrow and I will also add test for that case.

Thanks for catching that!

Reply:
Unfortunately bad news. As I implemented a path of migration 1.2 -> 1.3 -> 1.4. I found out we broke it with 1.4 version (I mean it is impossible to migrate job that was first running with 1.2).

It is due to the removal of NFA#Serializer class, which was used to serialize state in 1.2. We do not longer read or write anything there in 1.3+, but unfortunately the serializer is needed during the initialization. Will open a separate JIRA for that.

As for this PR, I think if we want to restore the migration path we will need to drop those changes.

Reply:
I am not sure if we should bring back these classes. 
I would recommend to simply throw an exception saying that 1.2 version is no longer supported. The reason for this is:
1) not many people seem to be using CEP in 1.2
2) the code is already a bit complicated with all the migration magic.

What do you think about that as a solution @dawidwys and @pnowojski ?

Reply:
If you @kl0u think that's the best solution I will agree :)

The only thing on which I would insist, is to extend `CEPMigrationTest` from 1.3 to 1.4/1.5 so it would cover the `and/or/subtype conditions` - that way it will be obvious from the automation tests that there is a backward compatibility with 1.3.

Reply:
I am for dropping support for 1.2. The only problem is regarding the 3rd point in @kl0u list. It is not possible to migrate 1.2 -> 1.4. Neither directly nor through 1.3.2. (that's what I was trying to say in my previous comment).

I would be for dropping support for <= 1.2 completely, if it is ok with you

Reply:
@dawidwys Let's drop support for <= 1.2. This will simplify the codebase and if someone complains, we can always re-introduce the missing classes.

Reply:
@kl0u I've reverted to the previous state (with removed *FilterFunctions).

@pnowojski Tests that checked migration from 1.3 with those conditions where there (whenever we use both `subtype` + `where` it is converted into `And`), but I added savepoints for versions 1.4 and 1.5.

Reply:
Wouldn't it be better to have those conditions tested explicitly? And what about or/subtype conditions? Are they covered as well?

Reply:
Sorry for the late response. 

Actually not sure about the value of testing those conditions explicitly. In fact it would just test if they are serializable. Anyway they are not intended for explicit usage (they could be annotated with `@Internal`, but it should be done as a separate task IMHO). Plus subtype is covered in migration tests (the `subtype` function on pattern).

Reply:
Sorry for late response. I've added test that says explicitly it tests agains Subtype/And/Or conditions.

I have not constructed this objects explicitly, as I think they are not intended to use them like this, but rather use chain of `where` for AND, `subtype` for SubtypeCondition and `or` for OrCondition.